### PR TITLE
Optimize EOF by reading types on demand

### PR DESCRIPTION
**Pull request by elijahhampton**
Introduces `EOF1Header::get_type` function to read individual EOF
code types. This is fast procedure because we need to read at most 4
bytes and a know offset. By doing so we avoid allocating separated
vector to types.

Closes https://github.com/ethereum/evmone/issues/1003.

### CODE REVIEW

**Comment by elijahhampton** [comment]
This PR addresses: https://github.com/ethereum/evmone/issues/1003

File: lib/evmone/eof.cpp [hunk 1]
```diff
@@ -79,6 +79,12 @@ EOFValidationError get_section_missing_error(uint8_t section_id) noexcept
     }
 }
 
+EOFCodeType get_type(bytes_view container, size_t header_size, size_t index)
+{
+    const size_t offset = header_size + index * 4;
+    return {container[offset], container[offset + 1], read_uint16_be(&container[offset + 2])};
```
Comment (chfast) [inline lib/evmone/eof.cpp:85]: This is good way of reading the `EOFCodeType`.

File: lib/evmone/eof.cpp [hunk 2]
```diff
@@ -220,14 +226,14 @@ std::variant<std::vector<EOFCodeType>, EOFValidationError> validate_types(
...
 
     // guaranteed by EOF headers validation
     assert(header_size + type_section_size < container.size());
 
-    for (auto offset = header_size; offset < header_size + type_section_size; offset += 4)
+    for (size_t i = 0; i < type_section_size / 4; ++i)
     {
-        types.emplace_back(
-            container[offset], container[offset + 1], read_uint16_be(&container[offset + 2]));
+        types.emplace_back(get_type(container, header_size, i));
```
Comment (chfast) [inline lib/evmone/eof.cpp:236]: However, this is not what was intended. We want to drop `std::vector<EOFCodeType> types`. And we want to replace all uses of  `header.types[fid]` with `header.get_type(container, index)`.

Hint: you will have to pass the container to the `EOF1Header::get_type()` because the header object don't have the reference to the container. We may change this later, but that's separate topic.

**Review by chfast — CHANGES_REQUESTED** [review]

File: lib/evmone/eof.cpp (hunk 2 above)
Comment (elijahhampton) [reply lib/evmone/eof.cpp:236]: Originally I had the method in the header class passing in the container, but I re read the issue and tried to mimic the suggested function signature.

I'll make the change

Comment (chfast) [reply lib/evmone/eof.cpp:236]: Sorry for the confusion then :)

Comment (elijahhampton) [reply lib/evmone/eof.cpp]: @chfast Fixed it

File: lib/evmone/eof.hpp [hunk 3]
```diff
@@ -15,6 +15,7 @@
 
 namespace evmone
 {
+inline uint16_t read_uint16_be(auto it) noexcept;
```
Comment (chfast) [inline lib/evmone/eof.hpp:18]: Move both of these helper functions here (`read_int16_be` and `read_uint16_be`).

File: lib/evmone/eof.hpp [hunk 4]
```diff
@@ -61,8 +62,25 @@ struct EOF1Header
     std::vector<uint16_t> container_sizes;
     /// Offset of every container section start;
     std::vector<uint16_t> container_offsets;
+    /// Size of every type section.
+    uint16_t type_section_size = 0;
```
Comment (chfast) [inline lib/evmone/eof.hpp:66]: Put these field after `version` as the type section is the first one.

File: lib/evmone/instructions.hpp [hunk 5]
```diff
@@ -1101,9 +1101,9 @@ inline code_iterator callf(StackTop stack, ExecutionState& state, code_iterator
     const auto index = read_uint16_be(&pos[1]);
     const auto& header = state.analysis.baseline->eof_header();
     const auto stack_size = &stack.top() - state.stack_space.bottom();
+    const EOFCodeType code_type = header.get_type(state.original_code, index);
```
Comment (chfast) [inline lib/evmone/instructions.hpp:1104]:
```suggestion
    const auto callee_type = header.get_type(state.original_code, index);
```

File: lib/evmone/eof.cpp [hunk 6]
```diff
@@ -254,6 +253,7 @@ struct InstructionValidationResult
     std::unordered_set<uint16_t> accessed_code_sections;
 };
 
+
```
Comment (chfast) [inline lib/evmone/eof.cpp:256]: Avoid adding new empty lines in unrelated code.

File: lib/evmone/eof.cpp [hunk 7]
```diff
@@ -286,9 +287,11 @@ std::variant<InstructionValidationResult, EOFValidationError> validate_instructi
         else if (op == OP_CALLF)
         {
             const auto fid = read_uint16_be(&code[i + 1]);
-            if (fid >= header.types.size())
+            if (fid >= header.type_section_size)
                 return EOFValidationError::invalid_code_section_index;
-            if (header.types[fid].outputs == NON_RETURNING_FUNCTION)
+
+            const EOFCodeType code_type = header.get_type(container, fid);
```
Comment (chfast) [inline lib/evmone/eof.cpp:293]:
```suggestion
            const auto type = header.get_type(container, fid);
```

File: lib/evmone/eof.cpp [hunk 8]
```diff
@@ -302,10 +305,12 @@ std::variant<InstructionValidationResult, EOFValidationError> validate_instructi
         else if (op == OP_JUMPF)
         {
             const auto fid = read_uint16_be(&code[i + 1]);
-            if (fid >= header.types.size())
+            if (fid >= header.type_section_size)
                 return EOFValidationError::invalid_code_section_index;
+
+            const EOFCodeType code_type = header.get_type(container, fid);
```
Comment (chfast) [inline lib/evmone/eof.cpp:311]:
```suggestion
            const auto type = header.get_type(container, fid);
```

File: lib/evmone/eof.cpp [hunk 9]
```diff
@@ -448,37 +455,41 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
         if (opcode == OP_CALLF)
         {
             const auto fid = read_uint16_be(&code[i + 1]);
+            if (fid >= header.get_type_count())
+                return EOFValidationError::invalid_code_section_index;
 
-            stack_height_required = code_types[fid].inputs;
+            const EOFCodeType code_type = header.get_type(container, fid);
```
Comment (chfast) [inline lib/evmone/eof.cpp:461]:
```suggestion
            const auto callee_type = header.get_type(container, fid);
```

File: lib/evmone/eof.cpp [hunk 10]
```diff
@@ -427,8 +433,9 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
 
     assert(!code.empty());
 
+    const EOFCodeType code_type_by_func_index = header.get_type(container, func_index);
```
Comment (chfast) [inline lib/evmone/eof.cpp:436]:
```suggestion
    const auto type = header.get_type(container, func_index);
```

File: lib/evmone/eof.cpp [hunk 11]
```diff
@@ -448,37 +455,41 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
         if (opcode == OP_CALLF)
         {
             const auto fid = read_uint16_be(&code[i + 1]);
+            if (fid >= header.get_type_count())
```
Comment (chfast) [inline lib/evmone/eof.cpp:458]: This is a new check. Why is it needed now?

Comment (elijahhampton) [reply lib/evmone/eof.cpp]: I removed the addition of those checks. Turns out the check I replaced in validate_instructions was not strict enough. Now it is and all test pass without the additional check.

File: lib/evmone/eof.hpp [hunk 12]
```diff
@@ -43,6 +61,11 @@ struct EOF1Header
     /// The EOF version, 0 means legacy code.
     uint8_t version = 0;
 
+    /// Size of every type section.
+    uint16_t type_section_size = 0;
+    /// Offset of type container section start.
```
Comment (chfast) [inline lib/evmone/eof.hpp:66]:
```suggestion
    /// Offset of type section start.
```

File: lib/evmone/eof.hpp [hunk 13]
```diff
@@ -43,6 +61,11 @@ struct EOF1Header
...
+    size_t type_section_offset = 0;
```
Comment (chfast) [inline lib/evmone/eof.hpp:67]: I think the offset field should be before the size field.

File: lib/evmone/eof.hpp [hunk 14]
```diff
@@ -62,7 +85,20 @@ struct EOF1Header
     /// Offset of every container section start;
     std::vector<uint16_t> container_offsets;
 
-    std::vector<EOFCodeType> types;
+    /// A helper to extract reference to a specific type section.
+    [[nodiscard]] EOFCodeType get_type(bytes_view container, size_t type_idx) const noexcept
+    {
+        static constexpr size_t TYPE_ENTRY_SIZE = 4;  // Size of each type entry in bytes
```
Comment (chfast) [inline lib/evmone/eof.hpp:91]:
```suggestion
        static constexpr size_t TYPE_ENTRY_SIZE = sizeof(EOFCodeType);  // Size of each type entry in bytes
```
Put this constant outside of function and use in other places that now uses `4`. E.g. `type_section_size != code_sizes.size() * 4`.

File: lib/evmone/eof.cpp [hunk 15]
```diff
@@ -582,6 +586,7 @@ std::variant<EOFValidationError, int32_t> validate_max_stack_height(
 
     const auto max_stack_height_it = std::max_element(stack_heights.begin(), stack_heights.end(),
         [](StackHeightRange lhs, StackHeightRange rhs) noexcept { return lhs.max < rhs.max; });
+
```
Comment (chfast) [inline lib/evmone/eof.cpp:589]: Unrelated empty line.

File: lib/evmone/eof.cpp [hunk 16]
```diff
@@ -759,14 +765,17 @@ std::variant<EOF1Header, EOFValidationError> validate_header(
...
-    if (const auto* error = std::get_if<EOFValidationError>(&types_or_error))
-        return *error;
-    const auto& types = std::get<std::vector<EOFCodeType>>(types_or_error);
+    const auto type_section_offset = header_size;
+    const auto type_section_size = section_headers[TYPE_SECTION].front();
+
+    if (type_section_size != code_sizes.size() * 4)
+        return EOFValidationError::invalid_type_section_size;
+
+    auto validation_error = validate_types(container, type_section_offset, type_section_size);
```
Comment (chfast) [inline lib/evmone/eof.cpp:774]: This is quite an issue: `validate_type()` should not be used in `validate_header()` and invoke separately after `validate_header()`. This way we can just pass `EOF1Header` to `validate_types()` and use its `.get_type()` method.

But maybe this requires a separate PR. What do you think @gumb0?

File: lib/evmone/eof.hpp [hunk 17]
```diff
@@ -43,6 +61,11 @@ struct EOF1Header
     /// The EOF version, 0 means legacy code.
     uint8_t version = 0;
 
+    /// Size of every type section.
+    uint16_t type_section_size = 0;
```
Comment (gumb0) [inline lib/evmone/eof.hpp:65]: We don't need to store the size at all, because header validation guarantees that `type_section_size == code_sizes.siz() * 4`

**Review by chfast — APPROVED** [review]
